src/codelabs/null-safety.md

```diff
@@ -0,0 +1,324 @@
+}
+```
+
+## Type promotion
+
+With sound null safety, Dart's
+[flow analysis](/null-safety/understanding-null-safety#flow-analysis) has
+been extended to take nullability into account. Nullable variables that
+can't possibly contain null values are treated like non-nullable variables.
+This behavior is called
+[type promotion](/null-safety/understanding-null-safety#type-promotion-on-null-checks).
+
+### Exercise: Definite assignment
+
+Dart's type system can track where variables are assigned and
+read, and can verify that non-nullable fields are given
+values before any code tries to read from them. This process is called
+[definite assignment](/null-safety/understanding-null-safety#definite-assignment-analysis).
+
+Try uncommenting the `if`-`else` statement in the code below, and
+watch the analyzer errors disappear:
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/definite_assignment.dart" replace="/if/\/\/if/g; /\ \ text\ =/\/\/  text =/g; /\ \ \}/  \/\/}/g"?>
+```dart:run-dartpad:ga_id-definite_assignment:null_safety-true
+void main() {
+  String text;
+
+  //if (DateTime.now().hour < 12) {
+  //  text = "It's morning! Let's make aloo paratha!";
+  //} else {
+  //  text = "It's afternoon! Let's make biryani!";
+  //}
+
+  print(text);
+  print(text.length);
+}
+```
+
+### Exercise: Null checking
+
+In the code below, add an `if` statement to the beginning of `getLength` that
+returns zero if `str` is null:
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/type_promotion.dart" replace="/.*if\ \(.*\n.*\n.*//g"?>
+```dart:run-dartpad:ga_id-null_checking:null_safety-true
+int getLength(String? str) {
+  // Add null check here
+
+  return str.length;
+}
+
+void main() {
+  print(getLength('This is a string!'));
+}
+```
+
+### Exercise: Promotion with exceptions
+
+Promotion works with exceptions as well as return statements. Try a null check
+that throws an `Exception` instead of returning zero.
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/promotion_exceptions.dart" replace="/.*if\ \(.*\n.*\n.*//g"?>
+```dart:run-dartpad:ga_id-promotion_exceptions:null_safety-true
+int getLength(String? str) {
+  // Try throwing an exception here if `str` is null.
+
+  return str.length;
+}
+
+void main() {
+  print(getLength(null));
+}
+```
+
+## The late keyword
+
+Sometimes variables — fields in a class, or top-level variables — 
+should be non-nullable, but they can't be
+assigned a value immediately. For cases like that, use the
+[`late` keyword]( /null-safety/understanding-null-safety#late-variables).
+
+When you put `late` in front of a variable declaration,
+that tells Dart the following:
+
+-   Don't assign that variable a value yet.
+-   You *will* assign it a value later.
+-   You'll make sure that the variable has a value *before* the variable is used.
+
+If you declare a variable `late` and the variable is read before it's assigned a
+value, a `LateInitializationError` is thrown.
```

Comment:
Unfortunately this is no longer a guarantee. This error has been removed from the SDK and the only behavior that is defined is that "some" error will be thrown. Refer to https://github.com/dart-lang/sdk/issues/44361#issuecomment-737672074 for more information.

```suggestion
value, an error is thrown by Dart.
```

Reply:
I removed "by Dart" because if you're going to use the passive voice, you should go passive all the way. :) (We should either say "an error is thrown" or "Dart throws an error", which isn't much more precise.)